### ISSUE TITLE
csf/lfd vs hackcheck, hackftp and others

### ISSUE THREAD

**Issue by zanami** [omega8cc/boa#1440]
Please explain what's the purpose of /var/xdrago/monitor/check/* scripts and how different are they from csf/lfd?

hackcheck seems to be doing similar thing to what LF_SSHD does but maybe less effective

For example, I have 43 entries in csf.deny (since guest-water.sh last run, see below) only 2 of them are not from lfd (they are from BOA.sh.txt).
hackcheck.archive.log contains only 3 lines.

Looks like this list only used by guest-water.sh (so once a day) to update allowed/denied lists via csf (so csf is a dependency).

Why both? What's the difference?

I'm asking because guest-water.sh clears the csf.deny list each time it runs. Maybe not completely but all those "permanently" banned by lfd IPs are allowed again.

Now to my story

I think in my case csf.deny gets flushed [here](https://github.com/omega8cc/boa/blob/master/aegir/tools/system/guest-water.sh#L326) because there's no hackftp.archive.x3.log (aslo no hackftp.archive.log and no ftp.log)
Running hackftp ends like this, no ftp.log file created

> perl hackftp
> 
> ===[]	GLOBAL===
> 
> CONTROL complete

> 
> ls /var/xdrago/monitor
> check/  hackcheck.archive.log  hackcheck.archive.x3.log  segfault_alert  ssh.log

Maybe everything is fine with ftp, no hack attempts etc. Not sure why it's the reason to completely flush csf.deny discarding what's lfd added to it in the past 24 hrs.

**Comment by zanami** [comment]
Ok I understand. 

Can you comment on hackftp/guest-water part? Why absence of hackftp.archive.x3.log triggers csf  -df while other logs don’t matter that much?
Will creating an empty /var/xdrago/monitor/hackcheck.archive.x3.log help me keep csf.deny intact?

**Comment by petrowsky** [comment]
I'm a bit interested in this too. I did some log parsing and my "friends" in China and Korea love to just pound on my boxes. Just from today, I've had over 1000 connection attempts. Here's just those when **Disconnected from authenticating...**.

```
root@s3:~# cat /var/log/auth.log | grep sshd | grep 'Disconnected from auth' | grep 'Oct 24' | wc -l
970
```

The result was **74** uniques from those **970** it found in auth.log.

```
root@s3:~# cat /var/log/auth.log | grep sshd | grep 'Disconnected from auth' | grep 'Oct 24' | cut -f11 -d ' ' | sort | uniq | wc -l
74
```

When look at my csf.deny there were no matching IP addresses in there from the various attempts from multiple ip addresses. How are the brute forcers being blocked?

Have you considered running something like sshguard in addition to your own stuff?

**Comment by petrowsky** [comment]
Running the following doesn't look like the temporary bans are working. This is the same on all three of my boxes.

```
root@s3:~# csf -t
csf: There are no temporary IP entries
```

Shouldn't I see entries with that many attempts on a single day within auth.log?

**Comment by petrowsky** [comment]
It looks like I clearly have persistent attempts which are not being blocked. These guys are just sitting on my doorstep.

```
root@s3:~# netstat -tuna | grep :22 | grep -v LISTEN
tcp        0      0 45.xx.xx.xx:22         211.44.226.158:48728    SYN_RECV
tcp        0      0 45.xx.xx.xx:22         112.175.127.179:48822   SYN_RECV
tcp        0      0 45.xx.xx.xx:22         112.175.127.179:40173   SYN_RECV
tcp        0      0 45.xx.xx.xx:22         112.175.127.179:43640   SYN_RECV
tcp        0      0 45.xx.xx.xx:22         211.44.226.158:61343    SYN_RECV
tcp        0      0 45.xx.xx.xx:22         211.44.226.158:52522    SYN_RECV
tcp        0     36 45.xx.xx.xx:22         47.xxx.xxx.xxx:49835    ESTABLISHED
tcp        0      0 45.xx.xx.xx:22         112.175.127.179:61794   SYN_RECV
tcp        0      0 45.xx.xx.xx:22         211.44.226.158:46133    SYN_RECV
tcp        0      0 45.xx.xx.xx:22         112.175.127.179:63896   SYN_RECV
tcp        0      0 45.xx.xx.xx:22         211.44.226.158:52263    SYN_RECV
tcp        0      0 45.xx.xx.xx:22         211.44.226.158:64825    SYN_RECV
tcp        0      0 45.xx.xx.xx:22         211.44.226.158:57085    SYN_RECV
```

For example if I take a look at Mr. 112.175.127.179 from Korea I get the following.

```
root@s3:~# grep 112.175.127.179 /var/log/syslog | wc -l
108
```

So where's the fail point. Why are these attempts not being blocked? What should I check for?

**Comment by petrowsky** [comment]
So, here's an image where I ran a split screen and used watch on the /run directory (ordered by time) watching for fire.pid.

On this particular box, despite it being created in April of this year and having just ran barracuda up-head today I never see fire.pid as expected from guest-fire.sh. It shows as being executed by cron.

Is there something I'm missing about how things are working?

![no_fire_pid](https://user-images.githubusercontent.com/96553/67541639-88afd080-f69e-11e9-91b0-4ec12c3048fe.png)

**Comment by zanami** [comment]
@petrowsky I'm receiving 20+ emails from lfd daily - "blocked distributed sshd attack on account [root]"
check your /etc/csf/csf.deny it's the place where ldf put them
check if lfd is running ( service lfd status)

and then there's /var/xdrago/guest-water.sh that can possibly flush csf.deny
— [if there's no /var/xdrago/monitor/hackftp.archive.x3.log file](https://github.com/omega8cc/boa/blob/master/aegir/tools/system/guest-water.sh#L322)
run _touch /var/xdrago/monitor/hackftp.archive.x3.log_ to fix this
— if there's no /root/.whitelist.site24x7.cnf (just once as it's created afterwards )

also /var/xdrago/guest-fire.sh
— [flushes deny lists](https://github.com/omega8cc/boa/blob/a635bc082c8b9b314d2b3892094b40961084e64e/aegir/tools/system/guest-fire.sh#L6) (both temp and permanent) if there are too many csf or guest-fire processes

I had to deal with both scripts to stop them from flushing csf.deny

I wouldn't rely on BOA monitor (hack* scripts) as it's relatively lame so check how your lfd works (or why it doesn't). Then check if the scripts above mess with your csf.deny

> I never see fire.pid as expected from guest-fire.sh. It shows as being executed by cron.

It only [writes fire.pid](https://github.com/omega8cc/boa/blob/a635bc082c8b9b314d2b3892094b40961084e64e/aegir/tools/system/guest-fire.sh#L28) when it does something i.e. /var/xdrago/monitor/ssh|web|ftp.log are not empty
otherwise guest_guard() executes in under a second and the script itself most of the time just does "sleep 5" with no pid present.

**Comment by petrowsky** [comment]
Yes, in fact, I can see that [minute.sh](https://github.com/omega8cc/boa/blob/d783f00394419d10a8ecdb553686feb035c3124e/aegir/tools/system/minute.sh#L493) runs hackcheck, yet hackcheck will [never add an IP](https://github.com/omega8cc/boa/blob/d783f00394419d10a8ecdb553686feb035c3124e/aegir/tools/system/monitor/check/hackcheck#L88) if guest-fire.sh exists.

I'm just trying to become familiar with the current system and figure out why IP's are not getting blocked.

**Comment by zanami** [comment]
> Yes, in fact, I can see that minute.sh runs hackcheck, yet hackcheck will never add an IP if guest-fire.sh exists.

It doesn't directly but it adds IPs to ssh.log and then guest-fire.sh parses it and runs csf -td
In my case all of these IPs would be already banned by lfd (permanently).
Why? What's these hack* scripts are for then? That's what I'm trying to figure out.

**Comment by omega8cc** [comment]
This regression has been fixed in https://github.com/omega8cc/boa/commit/e1e9252e8618e6c1ccf0014d2bb46cfb14e1a3d4

**Comment by omega8cc** [comment]
To quote the commit comment: "This has been added in response to non-fuctional csf/lfd systems with permanent entries (the ‘do not delete’ lines) exceeding the DENY_IP_LIMIT, which resulted with inability to ban newer IPs, while most of the older IPs were no longer an issue anyway. We should implement it so it will check the size of the ‘do not delete’ list instead of flushing it blindly, and then only once, not every 24 hours or hourly!" -- please let us know if you have any further questions. Thanks for bringing this regression to our attention. It's one of these things we don't use ourselves, because we don't use system/instance level firewall anyway, so it's not always easy to spot unless/until reported.

**Comment by zanami** [comment]
This looks strange to me
It should be triggered with 7 or more (fire.sh) processes if I’m not mistaken. According to this the next ps cannot find more than 6 (but it looks for guest-fire.sh)
Not sure why the first ps looks for just fire.sh and if it does make any difference. 

cat /var/log/fire-count.kill.log 
Wed Oct 23 21:25:01 +10 2019 Too many 6 guest-fire.sh processes killed
Wed Oct 30 18:23:01 +10 2019 Too many 5 guest-fire.sh processes killed
Thu Oct 31 17:39:02 +10 2019 Too many 5 guest-fire.sh processes killed
Thu Nov  7 03:47:01 +10 2019 Too many 3 guest-fire.sh processes killed
Thu Nov  7 16:50:01 +10 2019 Too many 5 guest-fire.sh processes killed
Thu Nov  7 20:26:01 +10 2019 Too many 4 guest-fire.sh processes killed
Fri Nov  8 05:27:01 +10 2019 Too many 4 guest-fire.sh processes killed
Fri Nov  8 10:59:01 +10 2019 Too many 4 guest-fire.sh processes killed
Fri Nov  8 19:22:01 +10 2019 Too many 4 guest-fire.sh processes killed
Sat Nov  9 13:59:01 +10 2019 Too many 5 guest-fire.sh processes killed
Sat Nov  9 18:17:01 +10 2019 Too many 4 guest-fire.sh processes killed
Sat Nov  9 21:54:01 +10 2019 Too many 4 guest-fire.sh processes killed

**Comment by omega8cc** [comment]
It shouldn’t matter and is probably a result of syncing the logic with host version designed to run on parent system with VServer. The problem is that it’s actually a bad idea to implement self monitoring and aggressive killing within the script itself. The csf_flood_guard() should be moved to some other script like minute.sh and separate monitoring script added for the host version.

**Comment by EdNett** [comment]
Hello,

We are running latest head 4.0.02 dev:

I'm not sure I shouldn't be opening a new issue for this or not, but a review of /etc/csf/csf.conf - shows that the CC_SRC=1 (which requires a maxmind license) and there is no maxmind license set by default. Shouldn't this be set to "2" if we are going to use CSF to block countries? 

Either this should be set to "2", or instructions shuold be given that one *MUST* obtain a maxmind license key since dec 2019.

**Comment by omega8cc** [comment]
Yes, it should be a separate issue.

**Comment by omega8cc** [comment]
This has been fixed a long time ago, closing.
